File: appium/Java/src/test/java/com/remotetestkit/appium/AndroidChromeTest.java
package com.remotetestkit.appium;

import java.io.File;
import java.net.URL;
import java.util.Arrays;

import io.appium.java_client.android.AndroidDriver;
import io.appium.java_client.android.nativekey.AndroidKey;
import io.appium.java_client.android.nativekey.KeyEvent;
import io.appium.java_client.android.options.UiAutomator2Options;
import org.apache.commons.io.FileUtils;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import org.openqa.selenium.OutputType;
import org.openqa.selenium.Keys;
import org.openqa.selenium.chrome.ChromeOptions;

public class AndroidChromeTest {

	private static AndroidDriver driver;

	@BeforeAll
	static void initAll() throws Exception {
		UiAutomator2Options capabilities = new UiAutomator2Options();

		// get userName, password from Environment variable
		capabilities.setCapability("userName", System.getenv("userName"));
		capabilities.setCapability("password", System.getenv("password"));
		capabilities.setCapability("platformName", "Android");
		capabilities.setCapability("deviceName", "Pixel");
		capabilities.setCapability("browserName", "Chrome");
		capabilities.setCapability("appiumVersion", "2.11.2");
		capabilities.setCapability("automationName", "UiAutomator2");

		driver = new AndroidDriver(new URL("https://gwjp.appkitbox.com/wd/hub"), capabilities);
	}

	@AfterAll
	static void tearDownAll() {
		driver.quit();
	}

	@Test
	void search() throws Exception {
		System.out.println("snapshotUrl: " + driver.getCapabilities().getCapability("snapshotUrl"));

		// Open URL
		String url = "https://www.google.com/";
		System.out.println("Open URL: " + url);
		driver.get(url);
		WebElement element = driver.findElement(By.name("q"));
		Thread.sleep(5000);
		FileUtils.copyFile(driver.getScreenshotAs(OutputType.FILE), new File("capture_B01.png"));

		// Input keys
		String word = "Remote TestKit";
		System.out.println("Input Keys: " + word);
		element.sendKeys(word);
		element.sendKeys(Keys.chord(Keys.ENTER));
		Thread.sleep(5000);
		FileUtils.copyFile(driver.getScreenshotAs(OutputType.FILE), new File("capture_B02.png"));

		// Get value
		String value = driver.findElement(By.name("q")).getAttribute("value");
		System.out.println("Text field value=" + value);
		Assertions.assertEquals(value, "Remote TestKit");

		driver.pressKey(new KeyEvent(AndroidKey.APP_SWITCH));
		driver.pressKey(new KeyEvent(AndroidKey.HOME));
	}
}

File: appium/proxy/Python/AndroidChromeTest.py
import os
import sys
import unittest
from time import sleep
from appium import webdriver
from selenium.webdriver.common.by import By
from appium.options.common import AppiumOptions

# ------------------------------------------------------------------------------------------
# The original selenium.webdriver.remote.remote_connection is not implemented to access
# the endpoint via proxy. So We wrote a monkey patch for the proxy.
import certifi
import urllib3
from selenium.webdriver.remote.remote_connection import RemoteConnection

RemoteConnection.__org__init__ = RemoteConnection.__init__


def patch_init(self, remote_server_addr, keep_alive=False, ignore_proxy=False):
    print("\nMonkey patch version: selenium.webdriver.remote.remote_connection")
    RemoteConnection.__org__init__(self, remote_server_addr, keep_alive=keep_alive, ignore_proxy=ignore_proxy)

    if keep_alive:
        # Define proxy. Default value is squid port.
        PROXY_URL = "http://localhost:3128"
        self._conn = urllib3.ProxyManager(proxy_url=PROXY_URL, timeout=self._timeout)

        # If basic authentication is required, uncomment the following and define it.
        # headers = urllib3.util.make_headers(proxy_basic_auth="userid:password")
        # self._conn = urllib3.ProxyManager(
        #     proxy_url=PROXY_URL, proxy_headers=headers, cert_reqs='CERT_REQUIRED', ca_certs=certifi.where())


RemoteConnection.__init__ = patch_init
# ------------------------------------------------------------------------------------------

# Specify userName, password to Environment variable
RTK_ACCESSTOKEN = os.environ.get('RTK_ACCESSTOKEN')

if not RTK_ACCESSTOKEN:
    print("Environment variable error")
    sys.exit()


class OpenUrlTest(unittest.TestCase):
    def setUp(self):
        caps = {
            'accessToken': RTK_ACCESSTOKEN,
            'deviceName': 'Nexus 5',
            'platformName': 'Android',
            'browserName': 'Chrome',
            'chromeOptions': {'w3c': False}
        }
        # Specify the endpoint
        options = AppiumOptions()
        for k in caps:
            options.set_capability(k, caps[k])
        self.driver = webdriver.Remote(
            "https://gwjp.appkitbox.com/wd/hub", options=options
        )
        print(f"command_executor={self.driver.command_executor}")
        print(f"proxy={self.driver.command_executor._conn.proxy}")

    def tearDown(self):
        self.driver.quit()

    def test_google_search(self):
        # print(self.driver.capabilities['snapshotUrl'])
        # Open URL
        url = "https://www.google.com/"
        print("Open URL: " + url)
        self.driver.get(url)
        element = self.driver.find_element(By.CSS_SELECTOR, 'textarea')
        sleep(5)
        self.driver.save_screenshot('capture_01.png')

        # Input keys
        word = "Remote testKit"
        print("Input Keys: " + word)
        element.send_keys(word)
        element.submit
        sleep(5)
        self.driver.save_screenshot('capture_02.png')

        # Get value
        value = self.driver.find_element(By.CSS_SELECTOR, 'textarea').get_attribute('value')
        print("Text field value=" + value)
        self.assertEqual(value, "Remote testKit")


if __name__ == '__main__':
    suite = unittest.TestLoader().loadTestsFromTestCase(OpenUrlTest)
    unittest.TextTestRunner(verbosity=2).run(suite)
